apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +37,120 @@ defmodule Yggdrasil.User do
+
+    permissions = query
+    |> Repo.all()
+    |> Enum.flat_map(fn ur -> ur.role.role_permissions end)
+    |> Enum.group_by(fn rp -> rp.resource end)
+    |> Enum.map(fn {r, rps} -> {String.to_atom(r), Enum.map(rps, fn rp -> String.to_atom(rp.permission) end)} end)
+    |> Enum.into(%{})
+
+    %{ user | permissions: permissions}
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted!(user, res_perms) do
+    case is_granted?(user, res_perms) do
+      {:ok, result}  -> result
+      {:error, msgs} -> raise RuntimeError, inspect(msgs)
+    end
+  end
+
+  def is_granted?(_user, []),  do: build_error ["resource keyword list is empty"]
+  def is_granted?(_user, nil), do: build_error ["resource keyword list is nil"]
+  def is_granted?(user, res_perms) do
+    res_perms
+    |> Enum.map(fn {res, perms} -> {res, user.permissions[res], perms} end)
+    |> Enum.map(&has_permissions/1)
+    |> Enum.reduce({:ok, true}, fn
+      {:ok, _},  acc={:ok, false}   -> acc
+      {:ok, res},    {:ok, true}    -> {:ok, res}
+      {:ok, _},  acc={:error, _}    -> acc
+      {:error, msg}, {:ok, _}       -> {:error, [msg]}
+      {:error, msg}, {:error, msgs} -> {:error, [msg|msgs]}
+    end)
+  end
+
+  defp has_permissions({_res, nil, _requried}), do: {:ok, false}
+  defp has_permissions({res, _given, nil}),     do: build_error "required permission list is nil for #{res}"
+  defp has_permissions({res, [], _required}),   do: build_error "given permission list is empty for #{res}"
+  defp has_permissions({res, _given, []}),      do: build_error "required permission list is empty for #{res}"
+  defp has_permissions({_res, given, required}) do
+    given = MapSet.new given
+    required = MapSet.new required
+
+    {:ok, MapSet.subset?(required, given)}
+  end
+
+  defp build_error(msg) do
+    Logger.error msg
+    {:error, msg}
+  end
```

Comment:
Ah, so I have avoided `:allowed|:denied` semantics because `is_granted?/2` does not make the decision about the resource it's merely confirming that the requested permissions exist and hands that back to the thing that will make a decision about the 'access' to the resource does that make sense?

For example the `EnsurePermission` plug would be a mechanism where that would fit perfectly it's taking the request, flows into the plug the plug checks the permissions and then either allows or denies passage. So `is_granted?/2` is really meant to be just building block and why I'm trying to keep it simple.

It would also be a bit awkward, because if you look at the character controller the plug is the first check to make sure you have the minimum to access the resource then in checks again within it to see if you have elevated permissions to augment the request so you would get something that reads like this
```elixir
# first check
plug EnsurePermission, [character: [:read]] when action in [:index]

# doing this inline here
# but user has the :read perm so we make it here
# now we want to see if they have the :all permission
# which elevates them to being able to see all characters
def index(conn, user, ...) do
  case User.is_granted?(user, [character: [:all]]) do
    :allowed -> # get all characters in system
    {:denied, _dont_care_here} -> # only the users characters
  end
end
```
By having those idioms in the return I have extra information in the call I don't need, and `:denied` seems glaring like I'm using this wrong while I'm clearly still allowing access even though it's denied.

All that said, I think you are right that with the `true|false` route it should just error and I think that is the route I want to go for now. It keeps this function simple without baking too many assumptions about why the permissions are being checked and can easily be composed into other things. I think that is idiomatic in this case, because the `errors` would be a case like this in delete/2: `If the model has no primary key, Ecto.NoPrimaryKeyFieldError will be raised.` the caller really has no option as to what to do about that error that makes sense.

So I think the trend is that `error` tuples are used in places where failures can occur and the caller can take an action that is sensible like accessing a file, it could very well not exist or fail to read etc where you may on error retry or let the user know and the select another. Perfect case for that. Same with db operations, unique constraint may fail because of some value being inserted and can be corrected etc. So error tuples work a treat, but as we see with delete there are some cases, which I have also noted in `Repo.one` where even though it's semantics are to return errors, certain cases are failures that get raised because there is nothing you could do really as the caller if say your model doesn't have a primary key. 

Anyway, sorry didn't mean to write so much but would you be ok with going forward for now with `is_granted?/2` with `true|false` and raise errors for those cases? We can always come back to it later but it may make more sense as we go along why I went this route once we have it use in more places. I also fully admit that this is completely uncharted territory for me, but I think it's inline with all that I see as far as raising the error in this case.